### PR TITLE
[Torch] Restore class-aware NMS for detection models by graph rewrite

### PR DESCRIPTION
**Pull request by masahi**
NMS used by PyTorch detection model actually performs multiclass NMS in one go, by adding different offsets to boxes from different classes so that two boxes from different classes never overlap. See 

https://github.com/pytorch/vision/blob/3d60f498e71ba63b428edb184c9ac38fa3737fa6/torchvision/ops/boxes.py#L80-L89

But this means most of O(N**2) IOU tests we do in the NMS triangle loop are useless. The goal of this PR is to restore class indices which is one of the inputs to `batched_nms` function above and perform class-aware NMS for TVM-compiled detection models. 

I did this by pattern matching and rewriting after model import. Specifically, I pattern match against [this subgraph](https://github.com/masahi/torchscript-to-tvm/blob/master/nms-perf-issue/maskrcnn_relay.txt#L1153-L1177) corresponding to PyTorch `batched_nms` used by maskrcnn / faster rcnn.

Unfortunately, this optimization didn't yield speedup I hoped: On GPU it only makes 70ms faster, and on CPU it actually makes it slightly slower (?) for some reason. I haven't looked into why it is not going much faster. 

nvprof output from running MaskRCNN on GPU

Before
```
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name                                                                                            
 GPU activities:   56.57%  711.34ms         2  355.67ms  15.296ms  696.05ms  fused_vision_non_max_suppression_kernel2                                                        
                   17.38%  218.54ms         1  218.54ms  218.54ms  218.54ms  fused_nn_dense_add_nn_relu_kernel0
```

After

```
            Type  Time(%)      Time     Calls       Avg       Min       Max  Name                                                                                            
 GPU activities:   54.52%  645.28ms         2  322.64ms  15.498ms  629.78ms  fused_vision_non_max_suppression_kernel2                                                        
                   18.44%  218.27ms         1  218.27ms  218.27ms  218.27ms  fused_nn_dense_add_nn_relu_kernel0     
```

On CPU, the output from VM profiler:

Before
```
#OpName                         #InvokeCount    #Duration(us): Sum/Mean/Min/Max
...
fused_vision_non_max_suppression        2               7902.54/3951.27/339.03/7563.51
```

After
```
#OpName                         #InvokeCount    #Duration(us): Sum/Mean/Min/Max
fused_vision_non_max_suppression        2               8129.3/4064.65/304.878/7824.42
```
So performance wise this change doesn't matter much, but I hope this also serves as a non trivial use of pattern matching and rewrite.

cc @kevinthesun @mbrookhart @zhiics @t-vi What do you think?

### CODE REVIEW

File: tests/python/frontend/pytorch/test_object_detection.py [hunk 1]
```diff
@@ -102,38 +105,55 @@ def test_detection_models():
...
+    num_tvm_valid_scores = count_valid_scores(tvm_scores)
 
     assert num_pt_valid_scores == num_tvm_valid_scores, (
         "Output mismatch: Under score threshold {}, Pytorch has {} valid "
         "boxes while TVM has {}.".format(score_threshold, num_pt_valid_scores, num_tvm_valid_scores)
     )
+
+    before = mod["main"]
+    after = rewrite(NMSRewrite(), before)
+    # TODO(masahi): Is there a better way to test if the desired rewrite has happened?
```
Comment (masahi) [inline tests/python/frontend/pytorch/test_object_detection.py:146]: @mbrookhart Any suggestion here?

**Comment by zhiics** [comment]
@masahi Thanks for the perf improvement. Could you provide the CPU numbers as well?

**Comment by masahi** [comment]
@zhiics Sure updated the description. The regression is only 200 us on CPU, so it may be just a measurement noise.

I have no idea why I'm not getting good speed up. IOU tests, including memory access to boxes should be definitely reduced. The only additional overhead I think of is that the input to NMS is one column wider, due to storing class ids. 

Performance is not great, but I believe having access to class ids should not be bad idea...

**Comment by zhiics** [comment]
@masahi I think this is an plausible as well particularly it is only in the parser. @kevinthesun please help take a look as well. Thanks.

**Comment by masahi** [comment]
I should mention that this rewrite is not run by default, so there is no perf risk.

File: python/tvm/relay/frontend/pytorch_utils.py [hunk 2]
```diff
@@ -25,3 +35,98 @@ def is_version_greater_than(ver):
     return "".join(re.findall(r"(\d+\.)(\d+\.)(\d)", torch.__version__)[0]) > "".join(
         re.findall(r"(\d+\.)(\d+\.)(\d)", ver)[0]
     )
+
+
+def batched_nms_pattern(boxes, scores, idxs, iou_threshold):
+    """A pattern to detect batched_nms function in torchvision"""
```
Comment (kevinthesun) [inline python/tvm/relay/frontend/pytorch_utils.py:41]: Can we have more comments about this pattern matching processing? I'm a bit confused of how class id is restored.

Comment (masahi) [reply python/tvm/relay/frontend/pytorch_utils.py:41]: Ok I'll update after the CI finishes. In the mean time, you can have a look at rewrite test cases in https://github.com/apache/tvm/blob/main/tests/python/relay/test_dataflow_pattern.py#L698-L736 to get an idea of how they work

File: tests/python/frontend/pytorch/test_object_detection.py (hunk 1 above)
Comment (mbrookhart) [reply tests/python/frontend/pytorch/test_object_detection.py:146]: What I do in the tests it write a reference version of the graph I want as the output and then assert that they match, see for instance https://github.com/apache/tvm/blob/e51bcdd1ed99fc813454c68911f8032c852f48b4/tests/python/relay/test_dataflow_pattern.py#L801-L815

**Comment by mbrookhart** [comment]
This is a bit of a shot in the dark.

I wonder if we're memory access limited, and so that's why you don't see a performance improvement.

When we do the nested loop, we always have to check if the id of instance k matches the id of instance j. Since the input shape is (batch_size, num_anchors, features), and features = 6 here, I wouldn't be surprised if checking the instance of k ends up reading all of the features of k into registers, and that memory read is the expensive operation. Once it's in memory, actually doing the iou calculation is relatively cheap, so skipping it doesn't help that much.

File: python/tvm/relay/frontend/pytorch_utils.py (hunk 2 above)
Comment (mbrookhart) [reply python/tvm/relay/frontend/pytorch_utils.py:41]: I think I understand what the rewritten pattern does, but I'm not understanding how the pytorch version works.

File: tests/python/frontend/pytorch/test_object_detection.py (hunk 1 above)
Comment (masahi) [reply tests/python/frontend/pytorch/test_object_detection.py:146]: Yes, the problem is the model is huge so manually creating the reference is not possible. Programmatically creating a reference requires the same pattern match & rewrite I want to test :)

File: python/tvm/relay/frontend/pytorch_utils.py (hunk 2 above)
Comment (masahi) [reply python/tvm/relay/frontend/pytorch_utils.py]: I added some comments to explain the pattern. It should be clear now.

File: tests/python/frontend/pytorch/test_object_detection.py (hunk 1 above)
Comment (mbrookhart) [reply tests/python/frontend/pytorch/test_object_detection.py:146]: I think this is fine for an integration test. If you want to write a unit test for your rewrite, I would suggest running it on a minimal example. I don't think it's strictly necessary.

What would you imagine a search and not-rewrite should return? You could test the pattern serach part by adding a throw into your callback and check that the raised error matches.

Comment (masahi) [reply tests/python/frontend/pytorch/test_object_detection.py]: > I think this is fine for an integration test. If you want to write a unit test for your rewrite, I would suggest running it on a minimal example. I don't think it's strictly necessary.

Ok that makes sense. I'm fine with what I have now. Removed TODO above.

I imagine search and not-rewrite would simply return a boolean and what you described would work as a quick implementation.

**Comment by masahi** [comment]
> When we do the nested loop, we always have to check if the id of instance k matches the id of instance j. Since the input shape is (batch_size, num_anchors, features), and features = 6 here, I wouldn't be surprised if checking the instance of k ends up reading all of the features of k into registers, and that memory read is the expensive operation. Once it's in memory, actually doing the iou calculation is relatively cheap, so skipping it doesn't help that much.

That's highly possible. Looking at this if condition in the triangle inner loop:
https://github.com/apache/tvm/blob/9956b5b859a24c8f4dec1abf308723b1257ffc66/python/tvm/topi/cuda/nms.py#L535-L540

previously, `force_suppress` is always True, so this condition short circuit and access to `out[base_idx + offset_k + id_index]` and `out[base_idx + offset_j + id_index]` just below never happen. But now, to make NMS class aware, I had to change `force_suppress` to False, and now access to `out[base_idx + offset_k + id_index]` and `out[base_idx + offset_j + id_index]` always happen. This may be cancelling the speedup from reduced IOU tests. Storing the class IDs in a different 1D tensor may help. 

That brings me to my pain point with our NMS API: I belieave our NMS API needs to be reworked. The current way of packing class ids and scores together with bbox coordinates is a design mistake that we inherited from MXNet. To store class ids, I have to cast ids to float32 and update and pass `id_index` appropriately. Since our NMS API also requires scores to be packed with bbox, all frontends except MXNet needs to do this concatenation. The worst part is that in NMS IR, the very first thing we do is the extraction 1D score tensor from packed data. So I see no good reason to pack score tensor and bbox coordinates.

**Comment by mbrookhart** [comment]
Sorry for the delay in responding to this, I wanted to look at the frameworks more closely. We currently have 5 importers that leverage NMS:

MXNET does multibox_transform_loc and then NMS on the outputs. multi_box_transform loc converts a 3D array of scores with shape (batch_size, class_num, num_anchors) into a most likely class and score for that class, plus does some coordinate transforms on the box

ONNX takes a 3D tensor of (batch_size, class, num_anchors), does slicing/concatenating with the boxes, and then does a per-class get_valid_counts->non_max

Pytorch takes in a 1D tensor of scores and concats it with the boxes before performing get_valid_counts and nms. As @masahi shows in this PR, there is preprocessing to embed all classes into that 1D tensor.

TF takes a 1D tensor of scores and concats it to the boxes before performing get_valid_counts and nms. I'm not sure if the rest of the TF graph is handling the loop over batch size and classes.

TFlite takes a 3D score tensor of shape (batch size, num_anchors, class_id), reorders it to (batch_size, class_id, num_anchors), performs multibox_transform_loc->nms, and strangely does get_valid_counts after NMS. 

It looks like we're doing pre-processing in every framework to reduce the amount of score information and convert it to the 5 or 6 D form the nms API wants, and then in at least TFLite and ONNX, we perform further spliting/slicing/concatenating to restore the separate class ids. 

I think I agree with @masahi, we seem to be jumping through a lot of hoops in the importers to support a TVM NMS API that's out of line with the frameworks, and that might be hurting our overall performance.

**Comment by trevor-m** [comment]
I highly agree with you guys.

For class-aware NMS, the [batch, num_anchors, 6] format seems very inefficient. It means all anchors need to be checked just to see if the classes match. A [batch, num_classes, num_anchors, 5] format would give us a nicely defined slice of memory where the same-class anchors are located.

> TF takes a 1D tensor of scores and concats it to the boxes before performing get_valid_counts and nms. I'm not sure if the rest of the TF graph is handling the loop over batch size and classes.

That's correct, TF's NMS is only for single class and single batch, so the TF graph loops over batches and classes. To do that, they use [tf.map_fn](https://www.tensorflow.org/api_docs/python/tf/map_fn) so the execution of each NMS can actually still run in parallel. However, this turns into a mess of control flow operators and TensorArrays, so Relay isn't able to do the same parallelization. This PR's graph rewrite could actually benefit TF OD models as well, but the pattern is a lot more complicated for TF.

**Comment by masahi** [comment]
@kevinthesun @zhiics @mbrookhart 

As shown in my new NMS PR https://github.com/apache/tvm/pull/7257, this rewrite results in a better speed up with improved memory layout. Can we merge this? I have newer rewrites coming to further optimize PyTorch NMS and MaskRCNN / FasterRCNN

**Review by mbrookhart — APPROVED** [review]
LGTM

**Review by zhiics — APPROVED** [review]

**Pull request merged**